File: 2251-Number of Flowers in Full Bloom.py
class Solution:
    def fullBloomFlowers(self, flowers: List[List[int]], people: List[int]) -> List[int]:
    # First try
        arr = []
        for start, end in flowers:
            arr.append((start, 1))
            arr.append((end+1, -1))
        
        arr.sort()
        preSum = [(0, 0)]
        curr = 0
        
        for val, flag in arr:
            if flag == 1:
                curr += 1
            else:
                curr -= 1
            preSum.append((val, curr))

        def bs(preSum, p):
            left, right = 0, len(preSum)-1
            while left + 1 < right:
                mid = left + (right-left)//2
                if preSum[mid][0]<=p:
                    left = mid
                else:
                    right = mid
            if preSum[right][0] <= p:
                return right
            return left
        
        res = []
        for p in people:
            idx = bs(preSum, p)
            res.append(preSum[idx][1])
            
        return res
        
    # Much better
        
        startArr, endArr = [], []
        for start, end in flowers:
            startArr.append(start)
            endArr.append(end+1)
        
        startArr.sort()
        endArr.sort()
        
        res = []
        
        for p in people:
            gain = bisect_right(startArr, p)
            loss = bisect_right(endArr, p)
            res.append(gain-loss)
            
        return res

File: C++/2251-Number of Flowers in Full Bloom.cpp
class Solution {
public:
    vector<int> fullBloomFlowers(vector<vector<int>>& flowers, vector<int>& people) {
        vector<int> gain;
        vector<int> loss;
        
        for (int i=0; i<flowers.size(); i++){
            gain.push_back(flowers[i][0]);
            loss.push_back(flowers[i][1]+1);
        }
        
        sort(gain.begin(), gain.end());
        sort(loss.begin(), loss.end());
        
        vector<int> res;
        for (int p:people){
            // Returns an iterator pointing to the first element in the range [first,last) which compares greater than val
            int add = upper_bound(gain.begin(), gain.end(), p) - gain.begin();
            int minus = upper_bound(loss.begin(), loss.end(), p) - loss.begin();
            res.push_back(add-minus);
        }
        
        return res;
    }
};